src/types/sgrStack.cpp

```diff
@@ -0,0 +1,240 @@
+    {
+        if (_numSavedAttrs > 0)
+        {
+            _numSavedAttrs--;
+
+            if (_nextPushIndex == 0)
+            {
+                _nextPushIndex = gsl::narrow<int>(_storedSgrAttributes.size() - 1);
+            }
+            else
+            {
+                _nextPushIndex--;
+            }
+
+            SavedSgrAttributes& restoreMe = _storedSgrAttributes.at(_nextPushIndex);
+
+            try
+            {
+                if (restoreMe.ValidParts.test(static_cast<size_t>(SgrSaveRestoreStackOptions::All)))
+                {
+                    return restoreMe.TextAttributes;
+                }
+                else
+                {
+                    return _CombineWithCurrentAttributes(currentAttributes,
+                                                         restoreMe.TextAttributes,
+                                                         restoreMe.ValidParts);
+                }
+            }
+            catch (std::out_of_range&)
+            {
+                // We should not be able to reach here: we pre-check that everything
+                // should be in range.
+                RaiseFailFastException(nullptr, nullptr, 0);
+            }
+        }
+
+        return currentAttributes;
+    }
+
+    TextAttribute SgrStack::_CombineWithCurrentAttributes(const TextAttribute& currentAttributes,
+                                                          const TextAttribute& savedAttribute,
+                                                          const AttrBitset validParts) // of savedAttribute
+    {
+        // If we are restoring all attributes, we should have just taken savedAttribute
+        // before we even got here.
+        FAIL_FAST_IF(validParts.test(static_cast<size_t>(SgrSaveRestoreStackOptions::All)));
+
+        TextAttribute result = currentAttributes;
+
+        // From xterm documentation:
+        //
+        //  CSI # {
+        //  CSI Ps ; Ps # {
+        //            Push video attributes onto stack (XTPUSHSGR), xterm.  The
+        //            optional parameters correspond to the SGR encoding for video
+        //            attributes, except for colors (which do not have a unique SGR
+        //            code):
+        //              Ps = 1  -> Bold.
+        //              Ps = 2  -> Faint.
+        //              Ps = 3  -> Italicized.
+        //              Ps = 4  -> Underlined.
+        //              Ps = 5  -> Blink.
+        //              Ps = 7  -> Inverse.
+        //              Ps = 8  -> Invisible.
+        //              Ps = 9  -> Crossed-out characters.
+        //              Ps = 2 1  -> Doubly-underlined.
+        //              Ps = 3 0  -> Foreground color.
+        //              Ps = 3 1  -> Background color.
+        //
+        //  (some closing braces for people with editors that get thrown off without them: }})
+        //
+        // Note that not all of these attributes are actually supported by

```

Comment:
Eh. We're about seventeen hours away from this being false now! When it was written, we didn't do Italic or Overline or Double or Invisible or Blink, but now we do. It's been a wild two years since you started this work (😢)

Reply:
Woot!